server/tracedb/azureappinsights.go

```diff
@@ -115,70 +115,181 @@ func (db *azureAppInsightsDB) GetTraceByID(ctx context.Context, traceID string)
 	return parseAzureAppInsightsTrace(traceID, table)
 }
 
-type columnIndex map[string]int
+type spanTable struct {
+	rows []spanRow
+}
+
+func (st *spanTable) Spans() []spanRow {
+	output := make([]spanRow, 0)
+	for _, row := range st.rows {
+		if row.Type() != "trace" {
+			output = append(output, row)
+		}
+	}
+
+	return output
+}
+
+func (st *spanTable) Events() []spanRow {
+	output := make([]spanRow, 0)
+	for _, row := range st.rows {
+		if row.Type() == "trace" {
+			output = append(output, row)
+		}
+	}
+
+	return output
+}
+
+type spanRow struct {
+	values map[string]any
+}
+
+func (sr *spanRow) Get(name string) any {
+	return sr.values[name]
+}
+
+func (sr *spanRow) Type() string {
+	return sr.values["itemType"].(string)
+}
+
+func (sr *spanRow) ParentID() string {
+	return sr.values["operation_ParentId"].(string)
+}
+
+func (sr *spanRow) SpanID() string {
+	return sr.values["id"].(string)
+}
+
+func newSpanTable(table *azquery.Table) spanTable {
+	spanRows := make([]spanRow, 0, len(table.Rows))
+	for _, row := range table.Rows {
+		spanRows = append(spanRows, newSpanRow(row, table.Columns))
+	}
+
+	return spanTable{spanRows}
+}
+
+func newSpanRow(row azquery.Row, columns []*azquery.Column) spanRow {
+	values := make(map[string]any)
+	for i, column := range columns {
+		name := *column.Name
+		if value := row[i]; value != nil {
+			values[name] = value
+		}
+	}
+
+	return spanRow{values}
+}
 
 func parseAzureAppInsightsTrace(traceID string, table *azquery.Table) (model.Trace, error) {
-	columnIndex := mapColumnNames(table.Columns)
-	spans := make([]model.Span, len(table.Rows))
+	spans, err := parseSpans(table)
+	if err != nil {
+		return model.Trace{}, err
+	}
+
+	return model.NewTrace(traceID, spans), nil
+}
 
-	for i, row := range table.Rows {
-		span, err := parseRowToSpan(row, columnIndex)
+func parseSpans(table *azquery.Table) ([]model.Span, error) {
+	spanTable := newSpanTable(table)
+	spanRows := spanTable.Spans()
+	eventRows := spanTable.Events()
+
+	spanEventsMap := make(map[string][]spanRow)
+	for _, eventRow := range eventRows {
+		spanEventsMap[eventRow.ParentID()] = append(spanEventsMap[eventRow.ParentID()], eventRow)
+	}
+
+	spanMap := make(map[string]*model.Span)
+	for _, spanRow := range spanRows {
+		span, err := parseRowToSpan(spanRow)
 		if err != nil {
-			return model.Trace{}, err
+			return []model.Span{}, err
 		}
 
-		spans[i] = span
+		spanMap[span.ID.String()] = &span
 	}
 
-	return model.NewTrace(traceID, spans), nil
+	for _, eventRow := range eventRows {
+		parentSpan := spanMap[eventRow.ParentID()]
+		event, err := parseEvent(eventRow)
+		if err != nil {
+			// don't fail conversion, just print a message
+			fmt.Printf(`failed to convert span event "%s" from span "%s" from AppInsights\n`, event.Name, eventRow.ParentID())
+		}
```

Comment:
I didn't want to fail the conversion here. The errors that return here are caused by parsing errors (timestamp and jJSON unmarshal). The reasoning is that I rather have fewer things in a span than not have any spans at all.